lib/galaxy/webapps/galaxy/api/job_files.py

```diff
@@ -71,54 +141,123 @@ def index(self, trans: ProvidesAppContext, job_id, **kwargs):
                 if any(jtid.dataset.dataset.purged for jtid in job.input_datasets):
                     raise exceptions.ItemDeletionException("Input dataset(s) for job have been purged.")
             else:
-                raise
+                raise HTTPException(status_code=404, detail="File not found.")
+        elif not os.path.isfile(path):
+            raise HTTPException(status_code=400, detail="Path does not refer to a file.")
+
+        return GalaxyFileResponse(path)
+
+    # The ARC remote job runner (`lib.galaxy.jobs.runners.pulsar.PulsarARCJobRunner`) expects a `PUT` endpoint to stage
+    # out result files back to Galaxy.
+    @router.put(
+        "/api/jobs/{job_id}/files",
+        summary="Populate an output file.",
+        responses={
+            201: {"description": "A new file has been created."},
+            204: {"description": "An existing file has been replaced."},
+            400: {"description": "Bad request (including no file contents provided)."},
+        },
+    )
+    async def populate(
```

Comment:
Sorry, no async possible, this is going to block the main event loop on IO.

Reply:
It looks we had the same thought at the same time :sweat_smile: https://github.com/galaxyproject/galaxy/pull/20098#discussion_r2063928989